src/main/java/dev/slimevr/osc/VRCOSCHandler.java

```diff
@@ -0,0 +1,252 @@
+	private static final long HMD_TIMEOUT = 15000;
+
+	public VRCOSCHandler(
+		HMDTracker hmd,
+		HumanPoseProcessor humanPoseProcessor,
+		WindowsNamedPipeBridge steamvrBridge,
+		OSCConfig oscConfig,
+		List<? extends ShareableTracker> shareableTrackers
+	) {
+		this.hmd = hmd;
+		this.humanPoseProcessor = humanPoseProcessor;
+		this.steamvrBridge = steamvrBridge;
+		this.config = oscConfig;
+		this.shareableTrackers = shareableTrackers;
+
+		trackersEnabled = new boolean[shareableTrackers.size()];
+
+		refreshSettings();
+	}
+
+	public void refreshSettings() {
+		// Sets which trackers are enabled and force HEAD to false
+		for (int i = 0; i < shareableTrackers.size(); i++) {
+			if (shareableTrackers.get(i).getTrackerRole() != TrackerRole.HEAD) {
+				trackersEnabled[i] = config
+					.getOSCTrackerRole(shareableTrackers.get(i).getTrackerRole(), false);
+			}
+		}
+
+		// Stops listening and closes OSC port
+		if (oscReceiver != null && oscReceiver.isListening()) {
+			oscReceiver.stopListening();
+		}
+		if (oscSender != null && oscSender.isConnected()) {
+			try {
+				oscSender.close();
+			} catch (IOException e) {
+				LogManager.warning("[VRCOSCHandler] Error closing the OSC sender: " + e);
+			}
+		}
+
+		if (config.getEnabled()) {
+			// Instantiates the OSC receiver
+			try {
+				int port = config.getPortIn();
+				oscReceiver = new OSCPortIn(
+					port
+				);
+				LogManager.info("[VRCOSCHandler] Listening to port " + port);
+			} catch (IOException e) {
+				LogManager
+					.warning("[VRCOSCHandler] Error listening to the port " + config.getPortIn());
+			}
+
+			// Starts listening for the Upright parameter from VRC
+			OSCMessageListener listener = this::handleReceivedMessage;
+			MessageSelector selector = new OSCPatternAddressMessageSelector(
+				"/avatar/parameters/Upright"
+			);
+			oscReceiver.getDispatcher().addListener(selector, listener);
+			oscReceiver.startListening();
+
+			// Instantiate the OSC sender
+			try {
+				InetAddress address = config.getAddress();
+				int port = config.getPortOut();
+				oscSender = new OSCPortOut(
+					address,
+					port
+				);
+				oscSender.connect();
+				LogManager
+					.info(
+						"[VRCOSCHandler] sending to port "
+							+ port
+							+ " at address "
+							+ address.toString()
+					);
+			} catch (IOException e) {
+				LogManager
+					.warning(
+						"[VRCOSCHandler] Error connecting to port "
+							+ config.getPortOut()
+							+ " at the address "
+							+ config.getAddress()
+					);
+			}
+		}
+	}
+
+	void handleReceivedMessage(OSCMessageEvent event) {
+		timeAtLastOSCMessageReceived = System.currentTimeMillis();
+		if (steamvrBridge != null && steamvrBridge.isConnected()) {
+			if (hmd.getStatus() != TrackerStatus.OK) {
+				hmd.setStatus(TrackerStatus.OK);
+			}
+			hmd.position
+				.set(
+					0f,
+					(float) event
+						.getMessage()
+						.getArguments()
+						.get(0) * humanPoseProcessor.getUserHeightFromConfig(),
```

Comment:
might want a more optimized way to do this but I’m too tired to figure it out

Reply:
this looks really weird, what on earth is going on here 🤔

Reply:
Getting VRC's Upright value (HMD'y position scaled to 0-1) and scaling it back to 0-userheight with proportions.

Reply:
Maybe SkeletonConfig could cache this... Otherwise I don't think there's a much better solution with the current setup of things

Reply:
caching it in SkeletonConfig and calling it from HumanPoseProcessor